test/e2e/far_e2e_test.go

```diff
@@ -0,0 +1,208 @@
+		ObjectMeta: metav1.ObjectMeta{Name: nodeName, Namespace: testNamespace},
+		Spec: v1alpha1.FenceAgentsRemediationSpec{
+			Agent:            agent,
+			SharedParameters: sharedparameters,
+			NodeParameters:   nodeparameters,
+		},
+	}
+	ExpectWithOffset(offsetExpect, k8sClient.Create(context.Background(), far)).ToNot(HaveOccurred())
+	return far
+}
+
+// deleteFAR delete the CR with offset
+func deleteFAR(far *v1alpha1.FenceAgentsRemediation) {
+	EventuallyWithOffset(offsetExpect, func() error {
+		err := k8sClient.Delete(context.Background(), far)
+		if apiErrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}, timeout, pollInterval).ShouldNot(HaveOccurred(), "failed to delete far")
+}
+
+// getKubeletReadyCondition return the  condition of Node kubelet, otherwise return an error
+func getKubeletReadyCondition(nodeName string) (corev1.NodeCondition, error) {
+	emptyCondition := corev1.NodeCondition{}
+	node, err := clientSet.CoreV1().Nodes().Get(context.Background(), nodeName, metav1.GetOptions{})
+	if err != nil {
+		return emptyCondition, err
+	}
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == corev1.NodeReady {
+			return condition, nil
+		}
+	}
+	return emptyCondition, fmt.Errorf("Node %s is not ready", nodeName)
+}
+
+// checkFarLogs get the FAR pod and check whether its logs has logString
+func checkFarLogs(logString string) {
+	By("checking logs")
+	var pod *corev1.Pod
+	EventuallyWithOffset(offsetExpect, func() *corev1.Pod {
+		pod = getFenceAgentsPod(testNamespace)
+		return pod
+	}, timeout, pollInterval).ShouldNot(BeNil(), "can't find the pod after timeout")
+
+	EventuallyWithOffset(offsetExpect, func() string {
+		logs, err := farUtils.GetLogs(clientSet, pod, "manager")
+		if err != nil {
+			log.Error(err, "failed to get logs. Might try again")
+			return ""
+		}
+		return logs
+	}, timeout/2, pollInterval).Should(ContainSubstring(logString))
+}
+
+// getFenceAgentsPod fetches the FAR pod based on FAR's label and namespace
+func getFenceAgentsPod(namespace string) *corev1.Pod {
+	pods := new(corev1.PodList)
+	podLabelsSelector, _ := metav1.LabelSelectorAsSelector(
+		&metav1.LabelSelector{MatchLabels: farController.FaPodLabels})
+	options := client.ListOptions{
+		LabelSelector: podLabelsSelector,
+		Namespace:     namespace,
+	}
+	if err := k8sClient.List(context.Background(), pods, &options); err != nil {
+		log.Error(err, "can't find the pod by it's labels")
+		return nil
+	}
+	if len(pods.Items) == 0 {
+		podNotFoundErr := &apiErrors.StatusError{ErrStatus: metav1.Status{
+			Status: metav1.StatusFailure,
+			Code:   http.StatusNotFound,
+			Reason: metav1.StatusReasonNotFound,
+		}}
+		log.Error(podNotFoundErr, "Zero containers for the pod")
+		return nil
+	}
+	return &pods.Items[0]
+}
+
+// wasNodeRebooted wait for kublet status to switch from flase to true
+func wasNodeRebooted(nodeName string, lastReadyTime metav1.Time) {
+	kubletPhase := "was ready"
```

Comment:
As posted earlier I don't think we can use the ready condition at all.
However, just as an exercise, I think this code is way too complicated.
Can't the test be reduced to `return cond.Status == corev1.ConditionTrue && cond.LastTransitionTime.After(lastReadyTime.Time)`? 🙂